### PR TITLE
[Refactor] Associative Embedding model

### PR DESCRIPTION
**Pull request by ly015**
## Motivation

Refactor Associative Embedding in 1.x

## Modification

In this PR:

- Add bottom-up pose estimator
- Add AE head and loss
- Add bottom-up image demo script
- Verified that the inference result is correct

To do in the future:

- Support bottom-up testing (need to modify the metric to work with bottom-up dataset)
- Support bottom-up training

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmpose/blob/master/.github/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit tests to ensure correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmpose/codecs/simcc_label.py [hunk 1]
```diff
@@ -105,15 +105,16 @@ def encode(
                 f'{self.smoothing_type}. Should be one of '
                 '{"gaussian", "standard"}')
 
-    def decode(self,
-               encoded: Tuple[np.ndarray,
-                              np.ndarray]) -> Tuple[np.ndarray, np.ndarray]:
+    def decode(self, simcc_x: np.ndarray,
```
Comment (ly015) [inline mmpose/codecs/simcc_label.py:108]: @Tau-J In this PR I slightly changed the decoding process in the base head. Consequently, `codec.decode()` takes packed arguments instead of a single tuple. Is that ok with you?

Comment (Tau-J) [reply mmpose/codecs/simcc_label.py:108]: No problem

File: configs/body_2d_keypoint/associative_embedding/coco/ae_hrnet-w32_8xb24-300e_coco-512x512.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            'pretrain_models/hrnet_w32-36af842e.pth'),
+    ),
+    head=dict(
+        type='AssociativeEmbeddingHead',
+        in_channels=32,
+        num_keypoints=17,
+        tag_dim=1,
+        tag_per_keypoint=True,
+        deconv_out_channels=None,
+        keypoint_loss=dict(type='KeypointMSELoss', use_target_weight=True),
```
Comment (jin-s13) [inline configs/body_2d_keypoint/associative_embedding/coco/ae_hrnet-w32_8xb24-300e_coco-512x512.py:91]: why not use `combined loss` here?

Comment (ly015) [reply configs/body_2d_keypoint/associative_embedding/coco/ae_hrnet-w32_8xb24-300e_coco-512x512.py:91]: AE uses exactly these 2 losses so I think it's better to explicitly include them in the argument list. CombinedLoss is more like a container (like ModuleDict), which is used in the `__init__()` to combine these 2 losses as one module (because the base head class assume there is one loss module).

File: demo/image_demo.py [hunk 3]
```diff
@@ -25,7 +25,9 @@ def parse_args():
     return args
 
 
-def main(args):
+def main():
```
Comment (Ben-Louis) [inline demo/image_demo.py:28]: Is this demo necessary after the bottom-up models have been supported?

File: mmpose/apis/inference.py [hunk 4]
```diff
@@ -176,14 +175,44 @@ def inference_topdown(model: nn.Module,
...
         results = []
+
+    return results
+
+
+def inference_bottomup(model: nn.Module, img: Union[np.ndarray, str]):
+    """Inference image with a bottom-up pose estimator.
+
+    Args:
+        model (nn.Module): The top-down pose estimator
```
Comment (Ben-Louis) [inline mmpose/apis/inference.py:195]: 'top-down' -> 'bottom-up'

File: demo/image_demo.py (hunk 3 above)
Comment (ly015) [reply demo/image_demo.py:28]: I think we can keep it for now in case someone wants to inference the model on cropped images.

**Review by Ben-Louis — APPROVED** [review]
